File: medium/909. Snakes and Ladders/909. Snakes and Ladders.ts
function snakesAndLadders(board: number[][]): number {
    const n = board.length
    board.reverse()

    const getAtPosition = (square: number): [number, number] => {
        const row = Math.floor((square - 1) / n)
        let column = (square - 1) % n

        if (row % 2)
            column = n - 1 - column

        return [row, column]
    }

    const q: number[][] = [[1, 0]] // [square index, steps]
    const visited = new Set()

    while (q.length) {
        let [square, steps] = q.shift()!

        for (let i = 1; i <= 6; i++) {
            let nextSquare = square + i
            const [row, column] = getAtPosition(nextSquare)

            if (board[row][column] !== -1)
                nextSquare = board[row][column]
            if (nextSquare === n ** 2) {
                return steps + 1
            }
            if (!visited.has(nextSquare)) {
                visited.add(nextSquare)
                q.push([nextSquare, steps + 1])
            }
        }
    }

    return -1
};

console.log(snakesAndLadders([[-1, -1, -1, -1, -1, -1], [-1, -1, -1, -1, -1, -1], [-1, -1, -1, -1, -1, -1], [-1, 35, -1, -1, 13, -1], [-1, -1, -1, -1, -1, -1], [-1, 15, -1, -1, -1, -1]]))
console.log(snakesAndLadders([[1, 1, -1], [1, 1, 1], [-1, 1, 1]]))

export { }

File: medium/909. Snakes and Ladders/909. Snakes and Ladders.go
package main

import (
	"math"
	"slices"
)

func getIntoPosition(square, n int) (int, int) {
	row := (square - 1) / n
	column := (square - 1) % n
	if row%2 == 1 {
		column = n - 1 - column
	}

	return row, column
}

func snakesAndLadders(board [][]int) int {
	n := len(board)
	target := math.Pow(float64(n), 2)
	slices.Reverse(board)

	queue := [][]int{{1, 0}}
	visited := make(map[int]bool)

	for len(queue) > 0 {
		item := queue[len(queue)-1]
		square, steps := item[0], item[1]
		queue = queue[1:]

		for i := 1; i < 7; i++ {
			nextSquare := square + i
			row, column := getIntoPosition(nextSquare, n)

			if board[row][column] != -1 {
				nextSquare = board[row][column]
			}
			if nextSquare == int(target) {
				return steps + 1
			}
			if visited[nextSquare] == false {
				visited[nextSquare] = true
				queue = append(queue, []int{nextSquare, steps + 1})
			}
		}
	}

	return -1
}
